File: services/dashboard/DashboardAnalytics/GraphQL/DatabaseTypes.cs
using DashboardAnalytics.Models;
using DashboardAnalytics.Models.DTOs;
using DashboardAnalytics.Services;
using HotChocolate.Subscriptions;
using System.Security.Claims;

namespace DashboardAnalytics.GraphQL;

public class DatabaseQuery
{
    public async Task<DatabaseOverviewResponse> GetDatabaseOverview(
        [Service] IDatabaseOverviewService service,
        bool forceRefresh = false,
        bool includeSampleDocuments = true)
    {
        var overview = await service.GetDatabaseOverviewAsync(forceRefresh);
        
        return new DatabaseOverviewResponse(
            overview.Id,
            overview.GeneratedAt,
            overview.Services.Select(MapToResponse).ToList(),
            MapStatsToResponse(overview.TotalStats),
            CacheTimeSeconds: 300
        );
    }

    public async Task<ServiceDatabaseResponse> GetServiceDatabase(
        string serviceName,
        [Service] IDatabaseOverviewService service,
        bool forceRefresh = false)
    {
        var serviceDb = await service.GetServiceDatabaseInfoAsync(serviceName, forceRefresh);
        return MapToResponse(serviceDb);
    }

    public async Task<List<DatabaseSearchResult>> SearchDatabases(
        SearchDatabaseRequest request,
        [Service] IDatabaseOverviewService service)
    {
        return await service.SearchDatabasesAsync(request);
    }

    public async Task<List<QueryExecutionHistoryResponse>> GetQueryHistory(
        [Service] IDatabaseOverviewService service,
        ClaimsPrincipal claimsPrincipal,
        int limit = 50,
        bool onlyMyQueries = false)
    {
        var userId = onlyMyQueries ? claimsPrincipal.FindFirst(ClaimTypes.NameIdentifier)?.Value : null;
        return await service.GetQueryHistoryAsync(userId, limit);
    }

    public async Task<List<DatabaseAlertResponse>> GetDatabaseAlerts(
        [Service] IDatabaseOverviewService service,
        bool includeResolved = false)
    {
        return await service.GetAlertsAsync(includeResolved);
    }

    private static ServiceDatabaseResponse MapToResponse(ServiceDatabase service)
    {
        return new ServiceDatabaseResponse(
            service.ServiceName,
            service.DatabaseName,
            service.ConnectionString,
            service.Port,
            service.Collections.Select(MapCollectionToResponse).ToList(),
            MapStatsToResponse(service.Stats),
            service.IsConnected,
            service.ErrorMessage
        );
    }

    private static CollectionInfoResponse MapCollectionToResponse(CollectionInfo collection)
    {
        return new CollectionInfoResponse(
            collection.Name,
            collection.DocumentCount,
            collection.SizeInBytes,
            collection.AverageSizeInBytes,
            collection.Indexes.Select(MapIndexToResponse).ToList(),
            collection.SampleDocument,
            collection.Schema
        );
    }

    private static IndexInfoResponse MapIndexToResponse(IndexInfo index)
    {
        return new IndexInfoResponse(
            index.Name,
            index.Keys,
            index.IsUnique,
            index.IsSparse,
            index.SizeInBytes
        );
    }

    private static DatabaseStatsResponse MapStatsToResponse(DatabaseStats stats)
    {
        return new DatabaseStatsResponse(
            stats.TotalCollections,
            stats.TotalDocuments,
            stats.TotalSizeInBytes,
            stats.TotalIndexes,
            stats.AverageDocumentSize
        );
    }
}

public class DatabaseMutation
{
    public async Task<QueryExecutionResponse> ExecuteQuery(
        ExecuteQueryRequest request,
        [Service] IDatabaseOverviewService service,
        ClaimsPrincipal claimsPrincipal)
    {
        var userId = claimsPrincipal.FindFirst(ClaimTypes.NameIdentifier)?.Value ?? "unknown";
        var userEmail = claimsPrincipal.FindFirst(ClaimTypes.Email)?.Value ?? "unknown";

        return await service.ExecuteQueryAsync(request, userId, userEmail);
    }

    public async Task<bool> RefreshDatabaseCache(
        [Service] IDatabaseOverviewService service,
        [Service] ITopicEventSender sender)
    {
        await service.ClearCacheAsync();
        var overview = await service.GetDatabaseOverviewAsync(forceRefresh: true);
        
        // Notify subscribers
        await sender.SendAsync("DatabaseRefreshed", new DatabaseUpdateEvent(
            ServiceName: "All",
            DatabaseName: "All",
            EventType: "CacheRefreshed",
            CollectionName: "All",
            Timestamp: DateTime.UtcNow,
            Metadata: null
        ));

        return true;
    }
}

public class DatabaseSubscription
{
    [Subscribe]
    [Topic("DatabaseUpdates")]
    public DatabaseUpdateEvent OnDatabaseUpdate(
        [EventMessage] DatabaseUpdateEvent updateEvent)
    {
        return updateEvent;
    }

    [Subscribe]
    [Topic("DatabaseRefreshed")]
    public DatabaseUpdateEvent OnDatabaseRefreshed(
        [EventMessage] DatabaseUpdateEvent updateEvent)
    {
        return updateEvent;
    }

    [Subscribe]
    [Topic("QueryExecuted")]
    public DatabaseUpdateEvent OnQueryExecuted(
        [EventMessage] DatabaseUpdateEvent updateEvent)
    {
        return updateEvent;
    }
}

File: frontend/src/types/database.types.ts
// Database Management Types
export interface DatabaseOverview {
  id: string;
  generatedAt: string;
  services: ServiceDatabase[];
  totalStats: DatabaseStats;
  cacheTimeSeconds: number;
}

export interface ServiceDatabase {
  serviceName: string;
  databaseName: string;
  connectionString: string;
  port: number;
  collections: CollectionInfo[];
  stats: DatabaseStats;
  isConnected: boolean;
  errorMessage?: string;
}

export interface CollectionInfo {
  name: string;
  documentCount: number;
  sizeInBytes: number;
  averageSizeInBytes: number;
  indexes: IndexInfo[];
  sampleDocument?: string;
  schema?: Record<string, string>;
}

export interface IndexInfo {
  name: string;
  keys: Record<string, number>;
  isUnique: boolean;
  isSparse: boolean;
  sizeInBytes: number;
}

export interface DatabaseStats {
  totalCollections: number;
  totalDocuments: number;
  totalSizeInBytes: number;
  totalIndexes: number;
  averageDocumentSize: number;
}

export interface QueryExecutionRequest {
  databaseName: string;
  collectionName: string;
  query: string;
  queryType: 'Find' | 'Aggregate' | 'Count';
  limit?: number;
  skip?: number;
}

export interface QueryExecutionResponse {
  id: string;
  isSuccessful: boolean;
  errorMessage?: string;
  results: string[];
  resultCount: number;
  executionTimeMs: number;
  executedAt: string;
}

export interface QueryExecutionHistory {
  id: string;
  userEmail: string;
  databaseName: string;
  collectionName: string;
  query: string;
  queryType: string;
  isSuccessful: boolean;
  resultCount: number;
  executionTimeMs: number;
  executedAt: string;
}

export interface DatabaseAlert {
  id: string;
  serviceName: string;
  databaseName: string;
  collectionName: string;
  alertType: string;
  message: string;
  severity: 'Info' | 'Warning' | 'Critical';
  metadata: Record<string, any>;
  isResolved: boolean;
  createdAt: string;
  resolvedAt?: string;
}

export interface SearchDatabaseRequest {
  searchTerm?: string;
  serviceName?: string;
  collectionName?: string;
  minDocumentCount?: number;
  maxDocumentCount?: number;
  minSizeInBytes?: number;
  maxSizeInBytes?: number;
}

export interface DatabaseSearchResult {
  serviceName: string;
  databaseName: string;
  collectionName: string;
  documentCount: number;
  sizeInBytes: number;
  matchedFields: string[];
}

export interface DatabaseUpdateEvent {
  serviceName: string;
  databaseName: string;
  eventType: string;
  collectionName: string;
  timestamp: string;
  metadata?: Record<string, any>;
}
